lib/puppet/ssl/host.rb

```diff
@@ -419,6 +429,140 @@ def load_crls(path)
+  def save_certificate_bundle(cert_bundle)
+    bundle_string = ''
+    cert_bundle.each do |cert|
+      bundle_string += cert.to_s
+    end
+    Puppet::Util.replace_file(certificate_location('ca'), 0644) do |f|
+      f.write(bundle_string)
+    end
+  end
+
+  # Attempts to load or fetch this host's certificate. Returns nil if
+  # no certificate could be found.
+  # @return Puppet::SSL::Certificate if found, nil otherwise
+  def get_host_certificate
+    if cert = check_for_certificate_in_memory(name)
+      return cert
+    elsif cert = check_for_certificate_on_disk(name)
+      return cert
+    elsif cert = download_certificate_from_ca(name)
+      save_host_certificate(cert)
+      return cert
+    else
+      return nil
+    end
+  end
+
+  # Checks the certificate indirection for a cert stored in memory.
+  # Only relevant if the memory terminus is in use, and currently
+  # only used in testing.
+  # @param [String] name the name of the cert to look for
+  # @return Puppet::SSL::Certificate or nil if none is found
+  def check_for_certificate_in_memory(name)
+    if Puppet::SSL::Certificate.indirection.terminus_class == :memory
+      return Puppet::SSL::Certificate.indirection.find(name)
+    end
+  end
+
+  # Checks for the requested certificate on disc, at a location
+  # determined by this host's configuration.
+  # @name [String] name the name of the cert to look for
+  # @return Puppet::SSL::Certificate or nil if none is found
+  def check_for_certificate_on_disk(name)
+    file_path = certificate_location(name)
+    if Puppet::FileSystem.exist?(file_path)
+      Puppet::SSL::Certificate.from_s(Puppet::FileSystem.read(file_path))
+    end
+  end
+
+  # Attempts to download this host's certificate from the CA server.
+  # Returns nil if the CA does not yet have a signed cert for this host.
+  # @param [String] name then name of the cert to fetch
+  # @return Puppet::SSL::Certificate or nil if none is found
+  def download_certificate_from_ca(name)
+    return nil if Puppet::SSL::Host.ca_location != :remote
+
+    response = Puppet::Rest::Routes.get_certificate(http_client, name)
+    if response.ok?
+      body = response.read_body
+      Puppet::SSL::Certificate.from_s(body)
+    end
+  end
+
+  # Saves the given certificate to disc, at a location determined by this
+  # host's configuration.
+  # @param [Puppet::SSL::Certificate] cert the cert to save
+  def save_host_certificate(cert)
+    file_path = certificate_location(name)
+    Puppet::Util.replace_file(file_path, 0644) do |f|
+      f.write(cert.to_s)
+    end
+  end
+
+  # Returns the file path for the named certificate, based on this host's
+  # configuration.
+  # @param [String] name the name of the cert to find
+  # @return [String] file path to the certs location
+  def certificate_location(name)
+    if Puppet::SSL::Host.ca_location == :only
+      name == 'ca' ? Puppet[:cacert] : File.join(Puppet[:signeddir], "#{name}.pem")
```

Comment:
Is it possible for our name to not be `'ca'` if the `ca_location` is `:only`?

Reply:
Yes. "Only" indicates that we only have access to the file structure of the CA, so certs with other names are stored in `:signeddir` instead of `:certdir`.

Reply:
`ca_location` is configured globally, on the Host class, not on instances of it. So there are places in the tests where we configure `Puppet::SSL::Host.ca_location = :only`, create our CA, then create new host objects to generate certificates representing agents. These also respect `ca_location = :only` and put their certs in `:signeddir`, where the CA then looks for them to see what it has signed. When I remove the logic about asking for the CA name, a bunch of tests fail in https://github.com/puppetlabs/puppet/blob/master/spec/integration/faces/ca_spec.rb, if you'd like to take a look.